### DISCUSSION TITLE
When can we expect attachCamera & attachAudio support for visionOS?

### DISCUSSION THREAD

**Discussion by Zigonja** [HaishinKit/HaishinKit.swift#1394] [Q&A]
Hello!

Currently attachCamera & attachAudio is not available in visionOS as it's blocked out by the macros in the code. Is it possible to just enable them or is there more to it as to why they're excluded?

**Comment by shogo4405** [comment]
I've heard that visionOS doesn't support camera or microphone input. At least, during the visionOS beta, there wasn't any API like https://developer.apple.com/documentation/avfoundation/avcapturesession available. That's why it's blocked out with macros.

Looking at it now, it seems to be enabled in the API documentation for vision OS 1.0+. However, it's unclear if it's actually usable.

**Reply by shogo4405** [reply]
It seems that some properties are unavailable, but AVCaptureSession itself appears to be implemented. Do you have a physical device? Are features like the camera and microphone accessible? If they are, I would like to make adjustments to make them available if possible.

<img width="503" alt="スクリーンショット 2024-03-27 16 03 57" src="https://github.com/shogo4405/HaishinKit.swift/assets/810189/082790c2-8d0e-4548-9dfe-65e0f6c05725">

**Reply by Zigonja** [reply]
Hey!

The physical camere for the persona feed is accessible, but only via, I did not try it with microphone yet but I suspect there is something simmilar.

[systemPreferredCamera](https://developer.apple.com/documentation/avfoundation/avcapturedevice/3955201-systempreferredcamera)

I was able to get the camera feed on my device as I do have it. I can gladly help you test the library if you dont own the device!

However the only way I was able to get the camera feed data was via

https://developer.apple.com/documentation/avfoundation/avcapturevideodataoutput/1389008-setsamplebufferdelegate

**Reply by shogo4405** [reply]
Sounds good! Since it's not available in Japan yet, I don't have the physical device.　I'm planning to modify the library to support attachCamera & attachAudio methods on visionOS. Please stay tuned!

If you can obtain CMSampleBuffer using the method described [here](https://developer.apple.com/documentation/avfoundation/avcapturevideodataoutput/1389008-setsamplebufferdelegate), you can try ingesting camera footage using the append method in [IOStream.append(sampleBuffer)](https://shogo4405.github.io/HaishinKit.swift/Classes/IOStream.html#/s:10HaishinKit8IOStreamC6appendyySo17CMSampleBufferRefaF)!

**Reply by Zigonja** [reply]
I can obtain it! I will try that as soon as I get to the device today/tromorrow and get back to you!

**Reply by shogo4405** [reply]
I have attempted to enable attachCamera. Unfortunately, it seems that AVCaptureAudioDataOutput is not provided, so attachAudio could not be opened. Please try https://github.com/shogo4405/HaishinKit.swift/pull/1395.

**Reply by Zigonja** [reply]
I think that on visionOS we can access the microphone through [AVAudioEngine](https://developer.apple.com/documentation/avfaudio/avaudioengine)

**Reply by shogo4405** [reply]
If AVAudioEngine is available, I've prepared an interface to send AVAudioBuffer output from AVAudioEngine like this.
- https://github.com/shogo4405/HaishinKit.swift/blob/main/Examples/iOS/AudioCapture.swift
- https://shogo4405.github.io/HaishinKit.swift/Classes/IOStream.html#/s:10HaishinKit8IOStreamC6append_4whenySo13AVAudioBufferC_So0F4TimeCtF

**Reply by devon3000** [reply]
I've tested this approach on VisionOS for RTMP Streaming and it works with 1.9.7, but I can't figure out to get it to work in 2.0.x because we need to use MediaMixer, which has a dependency on AVCaptureAudioDataOutput.

It would be very helpful if there was some fallback version of MediaMixer.attachAudio in 2.0.x which would use AVAudioEngine when compiling for visionOS, so we could get microphone audio on Vision Pro in 2.0.x implementations.

**Reply by shogo4405** [reply]
There is a similar method available in MediaMixer, and I intend for it to be used as well.
https://github.com/shogo4405/HaishinKit.swift/blob/main/HaishinKit/Sources/Mixer/MediaMixer.swift#L274-L281

Please try it.

**Reply by devon3000** [reply]
Thank you for the pointer! This was exactly what I needed.

I have published an example project that uses HaishinKit and the Apple Enterprise API to live-stream the Vision Pro front camera to YouTube.
https://github.com/devon3000/AVPEnterpriseAPI-CameraStreaming

Thank you very much @shogo4405 for your excellent library, this would have been much more difficult without it.

**Reply by shogo4405** [reply]
@devon3000 
That’s great! Please make sure to register and introduce it in a new Discussion. It might be helpful for others and could lead to collaboration!

https://github.com/shogo4405/HaishinKit.swift/discussions/new?category=show-and-tell
